### PR TITLE
test: add cache testing suite

### LINKED ISSUE #3852: Add cache support for responses without `cache-control: public`
According to https://www.rfc-editor.org/rfc/rfc9111.html#response.cacheability, it's possible to cache responses if `Cache-Control: public` is not present.

Going through that list (non exhaustively):

- [ ] Support caching data with `Cache-Control: private`, behind an option (disabled by default)
- [ ] Support `Expires` header
- [ ] Caching responses that are not explicitly uncacheable, behind an option (disabled by default)

On that latter point:

> Note that, in normal operation, some caches will not store a response that has neither a cache validator nor an explicit expiration time, as such responses are not usually useful to store. However, caches are not prohibited from storing such responses.

This might be useful for some application.

### LINKED ISSUE #3869: sql cache should always have a entry size limit
We can't store more than 2GB.

### PR DESCRIPTION
**Pull request by flakey5**
This adds the test suite from https://cache-tests.fyi to test the cache interceptor.

Note these aren't compliance tests, we don't need to pass 100% of these. Arguably we shouldn't since some of the tests aren't even applicable to Undici (i.e. cdn-specific, we can skip these though). However, I do think we should try to rely on these in CI.

Current issues:
* [x] #3658 needs to be merged
* [x] Lots of tests failing due to network errors (`TypeError: fetch failed`, mostly socket & http protocol issues)
* [ ] Some test failures

Closes #3852 
Closes #3869 

cc @mcollina @ronag

### CODE REVIEW

File: test/cache-interceptor/cache-tests.mjs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import { Client, interceptors, setGlobalDispatcher } from '../../index.js'
+import { getResults, runTests } from '../fixtures/cache-tests/test-engine/client/runner.mjs'
+import tests from '../fixtures/cache-tests/tests/index.mjs'
+import { once } from 'node:events'
+import { determineTestResult, testLookup } from '../fixtures/cache-tests/test-engine/lib/results.mjs'
+
+const PROTOCOL = 'http'
+const PORT = 8000
+const BASE_URL = `${PROTOCOL}://localhost:${PORT}`
+const PIDFILE = '/tmp/http-cache-test-server.pid'
```
Comment (KhafraDev) [inline test/cache-interceptor/cache-tests.mjs:15]: this won't work on windows

File: test/cache-interceptor/cache-tests.mjs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+// Wait for the server to start
+await once(worker, 'message')
+
+// Expose cache globally
+const client = new Client(BASE_URL).compose(interceptors.cache())
+setGlobalDispatcher(client)
+
+// Run the tests
+await runTests(tests, globalThis.fetch, false, BASE_URL).catch(err => {
```
Comment (KhafraDev) [inline test/cache-interceptor/cache-tests.mjs:35]: this should use undici's fetch, not the global

**Review by KhafraDev — COMMENTED** [review]
> Note these aren't compliance tests, we don't need to pass 100% of these. Arguably we shouldn't since some of the tests aren't even applicable to Undici.

If these tests are targeting fetch, and fetch is a browser spec, why wouldn't some of the tests be applicable?

**Comment by flakey5** [comment]
> If these tests are targeting fetch

They aren't targeting fetch. It has tests for every part of the stack that can implement caching. They're written in js w/ fetch so that they can be ran in browsers, but browsers aren't the only thing the tests cover. A lot of the behaviors do overlap, however, there are still things specific to servers or cdns. 

There are also tests for things that rfc9111 obsoletes (such as using the `pragma` header for caching directives)

**Comment by KhafraDev** [comment]
> They aren't targeting fetch

What I mean is that the tests explicitly use fetch and it tests browser caching behavior (every http request made through browsers is done through fetch, at least in the specs). Anything that doesn't work would seem like a bug to me.

> A lot of the behaviors do overlap, however, there are still things specific to servers or cdns.

Is there anything specific?

> There are also tests for things that rfc9111 obsoletes (such as using the pragma header for caching directives)

Is lack of pragma support intentional, because it's obsolete? Is there no plan on adding support for it?

**Comment by flakey5** [comment]
> Is there anything specific?

They're mostly tests for `CDN-Cache-Control`

> Is lack of pragma support intentional, because it's obsolete? Is there no plan on adding support for it?

It was intentionally left out, right now there aren't any plans to add it.

**Comment by flakey5** [comment]
So it does look like we can run tests that only apply to browsers, which should skip the cdn tests

**Comment by mcollina** [comment]
Why did you lift the entirety of the source code of that project? What's the long term maintainability strategy?

Either:

- we can use the npm module they publish, so just add a dependency
- we vendor all their repo, but you also add a way to automatically update it
- we only lift the test cases, essentially making future updates on these harder

Anyhow, I would recommend _not_ to use `fetch()` for this, because it'd make things harder long term. Use `undici.request()`.

**Comment by KhafraDev** [comment]
> we can use the npm module they publish, so just add a dependency

We can't - it hasn't been updated in 4 years and there seem to be relevant tests added and fixed in that time.

> Anyhow, I would recommend not to use fetch() for this, because it'd make things harder long term. Use undici.request().

afaict the tests are specifically vendored for use in browsers and the `runTests` function they export expects fetch. Using anything other than fetch would likely end up being harder to maintain.

**Comment by flakey5** [comment]
> Why did you lift the entirety of the source code of that project

In addition to what @KhafraDev said, there's a fix I needed to make to the source (re http-tests/cache-tests#139)

> afaict the tests are specifically vendored for use in browsers and the runTests function they export expects fetch. Using anything other than fetch would likely end up being harder to maintain.

+1

**Review by ronag — APPROVED** [review]

File: test/cache-interceptor/cache-tests.mjs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+          reject(new Error('worker thread startup timed out'))
+        }, 5000)
+      })
+    ])
+
+    const client = new Agent().compose(interceptors.cache(environment.opts))
+    setGlobalDispatcher(client)
+
+    // Run the tests
+    await runTestSuite(tests, fetch, true, BASE_URL)
```
Comment (KhafraDev) [inline test/cache-interceptor/cache-tests.mjs:214]: the test suite is still using the global fetch, not the one that undici exports.

File: test/cache-interceptor/cache-tests.mjs [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+
+  try {
+    await Promise.race([
+      once(worker, 'message'),
+      new Promise((_resolve, reject) => {
+        setTimeout(() => {
+          reject(new Error('worker thread startup timed out'))
+        }, 5000)
+      })
+    ])
```
Comment (KhafraDev) [inline test/cache-interceptor/cache-tests.mjs:208]:
```suggestion
    await once(worker, 'message', { signal: AbortSignal.timeout(5000) })
```

nit

**Review by KhafraDev — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**